rocketpool/api/pdao/set-snapshot-address.go

```diff
@@ -0,0 +1,130 @@
+package pdao
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/rocket-pool/rocketpool-go/network"
+	"github.com/rocket-pool/smartnode/shared/services"
+	"github.com/rocket-pool/smartnode/shared/types/api"
+	"github.com/rocket-pool/smartnode/shared/utils/eth1"
+	"github.com/urfave/cli"
+)
+
+func canSetSnapshotAddress(c *cli.Context, snapshotAddress common.Address, signature string) (*api.PDAOCanSetSnapshotAddressResponse, error) {
+
+	// Get services
+	if err := services.RequireNodeWallet(c); err != nil {
+		return nil, err
+	}
+	if err := services.RequireRocketStorage(c); err != nil {
+		return nil, err
+	}
+	w, err := services.GetWallet(c)
+	if err != nil {
+		return nil, err
+	}
+	rp, err := services.GetRocketPool(c)
+	if err != nil {
+		return nil, err
+	}
+	nodeAccount, err := w.GetNodeAccount()
+	if err != nil {
+		return nil, err
+	}
+
+	// Response
+	response := api.PDAOCanSetSnapshotAddressResponse{}
+
+	votingInitialized, err := network.GetVotingInitialized(rp, nodeAccount.Address, nil)
+	if !votingInitialized {
+		return nil, fmt.Errorf("Voting must be initialized to set a snapshot address. Use 'rocketpool pdao initialize-voting' to initialize voting first.")
+	}
+	// Get transactor
+	opts, err := w.GetNodeAccountTransactor()
+	if err != nil {
+		return nil, err
+	}
+
+	// Gas info
+	gasInfo, err := network.EstimateInitializeVotingGas(rp, opts)
+	if err != nil {
+		return nil, fmt.Errorf("Could not estimate the gas required to claim RPL: %w", err)
+	}
+	response.GasInfo = gasInfo
+
+	// Update response
+	return &response, nil
+}
+
+func setSnapshotAddress(c *cli.Context, snapshotAddress common.Address, signature string) (*api.PDAOSetSnapshotAddressResponse, error) {
+
+	// Get services
+	if err := services.RequireNodeWallet(c); err != nil {
+		return nil, err
+	}
+	if err := services.RequireRocketStorage(c); err != nil {
+		return nil, err
+	}
+	w, err := services.GetWallet(c)
+	if err != nil {
+		return nil, err
+	}
+	rp, err := services.GetRocketPool(c)
+	if err != nil {
+		return nil, err
+	}
+	nodeAccount, err := w.GetNodeAccount()
+	if err != nil {
+		return nil, err
+	}
+
+	// Response
+	response := api.PDAOSetSnapshotAddressResponse{}
+
+	votingInitialized, err := network.GetVotingInitialized(rp, nodeAccount.Address, nil)
+	if !votingInitialized {
+		return nil, fmt.Errorf("Voting must be initialized to set a snapshot address. Use 'rocketpool pdao initialize-voting' to initialize voting first.")
+	}
+
+	// Slice signature string into v, r, s component of a signature giving node permission to use the given signer
+	str_v := signature[len(signature)-2:]
+	str_r := signature[:64]
+	str_s := signature[64:128]
```

Comment:
can we add a function that expects a 129 byte 0x-prefixed EIP-712 signature and returns v/r/s as `v uint8` and `r, s []byte`?

And, ideally, a unit test :)

Reply:
I truncated the 0x prefix [here](https://github.com/rocket-pool/smartnode/blob/54da653252683cf4b36bc804f993d69cc42a76ba/shared/utils/cli/validation.go#L85). Your suggestion is def a better way to handle it though. Will do, thanks for the review :)